google/resource_compute_instance_group_manager.go

```diff
@@ -450,59 +575,36 @@ func resourceComputeInstanceGroupManagerUpdate(d *schema.ResourceData, meta inte
 			return err
 		}
 
-		if d.Get("update_strategy").(string) == "RESTART" {
-			managedInstances, err := config.clientComputeBeta.InstanceGroupManagers.ListManagedInstances(
-				project, zone, d.Id()).Do()
-			if err != nil {
-				return fmt.Errorf("Error getting instance group managers instances: %s", err)
-			}
-
-			managedInstanceCount := len(managedInstances.ManagedInstances)
-			instances := make([]string, managedInstanceCount)
-			for i, v := range managedInstances.ManagedInstances {
-				instances[i] = v.Instance
-			}
-
-			recreateInstances := &computeBeta.InstanceGroupManagersRecreateInstancesRequest{
-				Instances: instances,
-			}
+		err = performUpdate(config, d.Id(), d.Get("update_strategy").(string), d.Get("rolling_update_policy").([]interface{}), project, zone)
+		if err != nil {
+			return err
+		}
 
-			op, err = config.clientComputeBeta.InstanceGroupManagers.RecreateInstances(
-				project, zone, d.Id(), recreateInstances).Do()
-			if err != nil {
-				return fmt.Errorf("Error restarting instance group managers instances: %s", err)
-			}
+		d.SetPartial("instance_template")
+	}
 
-			// Wait for the operation to complete
-			err = computeSharedOperationWaitTime(config.clientCompute, op, project, managedInstanceCount*4, "Restarting InstanceGroupManagers instances")
-			if err != nil {
-				return err
-			}
+	// If version changes then update
+	if d.HasChange("version") {
+		manager := &computeBeta.InstanceGroupManager{
+			Versions: expandVersions(d.Get("version").([]interface{})),
 		}
 
-		if d.Get("update_strategy").(string) == "ROLLING_UPDATE" {
-			// UpdatePolicy is set for InstanceGroupManager on update only, because it is only relevant for `Patch` calls.
-			// Other tools(gcloud and UI) capable of executing the same `ROLLING UPDATE` call
-			// expect those values to be provided by user as part of the call
-			// or provide their own defaults without respecting what was previously set on UpdateManager.
-			// To follow the same logic, we provide policy values on relevant update change only.
-			manager := &computeBeta.InstanceGroupManager{
-				UpdatePolicy: expandUpdatePolicy(d.Get("rolling_update_policy").([]interface{})),
-			}
+		op, err := config.clientComputeBeta.InstanceGroupManagers.Patch(project, zone, d.Id(), manager).Do()
```

Comment:
I'm still not totally sure this will work. For rolling updates, wouldn't we want the UpdatePolicy in the same patch call that we're actually making the update in?

Reply:
It works because the first Patch (the one triggered if "version" field changes) updates versions and their parameters (target size and instance template) but does not trigger the update (as it is considered as an opportunistic update). The second Patch (called through performUpdate method) will trigger a proactive update.
I agree this is not ideal because there is a short window where the first opportunistic update could be triggered before the second proactive update.
Will take a look and see if I can refactor this part!